keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+    displayName: "zookeeper cluster URI"
+    description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+    default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: true
+```
+
+```yaml
+type: object
+"$schema": "http://json-schema.org/draft-07/schema#"
+apiVersion: kudo.dev/v1beta2
+properties:
+  Zookeeper:
+    properties:
+      URI:
+        type: "string"
+        description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+        default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: ["URI"]
+```
+
+KUDO will support both package formats for the future. At some point the old format might become deprecated, but this is not planned yet.
+
+Note: The apiVersions here (`v1beta1` and `v2beta2`) correspond to the CRD versions, but are separate definitions, even if they seem to be similar in certain aspects.
+
+### CRDs
+
+Normally, per Kubernetes conventions, CRDs would evolve over released versions without a breaking change from one version to the next, allowing clients to gradually convert to new CRD versions. This requires for a CRD to be convertible from the old version to new _and vice versa_ for all clients (new and old) to be able to read/write both old and new CRD versions. However, such a conversion in both directions will not be possible with this change, as JSON-schema supports a lot of configurations which will not be mappable from a flat list. 
+
+There will be a hard break for clients and users of CRDs from `v1beta1` to the new `v1beta2`. We will provide a migration path for existing installed CRDs, but the clients accessing these CRDs will need to be updated.
+
+#### Conversion
+
+To migrate the existing CRDs, we will implement a Conversion Webhook that only allows `v1beta1` to `v1beta2` conversion, not the other way around. When KUDO is upgraded, it will:
+
+1. Deploy the new CRD version where the new (`v1beta2`) version will be the _only_ served and stored version:
+    - `v1beta1` will be still in the list, but with `storage=false, served=false`
+    - `v1beta2` will be marked as `storage=true, served=true`
+2. Deploy the new manager with the conversion webhook
+3. Run Pre-Migration: fetch all existing CRs, run the conversion and check for any errors
+4. Run the migration: fetch all CRDs and issue an update - This will trigger a conversion and save the CRD with the new version
+5. With a later major KUDO release, we can remove `v1beta1` from the CRD list - this should be at least 6 month later, as soon as we remove support for the `v1beta1` version we cannot support migrating existing installations anymore.
+
+Reference: [CRD Versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/)
+
+The actual type that we are going to use in the CRD is `extensionapi.JSON` - This type wraps a `[]byte` and allows storage of any data.  
```

Comment:
I don't understand this comment, can you guys elaborate? I think I did not get how we got from very high level migration docs into types in go 🤔 I am missing the link.

Reply:
oh I see, it goes into more details in the OV description. Maybe drop it from here then? Or be more verbose

Reply:
Yeah, i'm not sure why it landed here. It's enough to have the details in the OV section.